### DISCUSSION TITLE
Migration to C#8 / C#9 - .NET 5?

### DISCUSSION THREAD

**Discussion by PizzaConsole** [oqtane/oqtane.framework#709] [General]
Would there be any benefits to migrating Oqtane to a newer version of C#?

**Comment by sbwalker** [comment]
C# 9 will be released with .NET 5 - so its something that should be considered once .NET 5 ships.

**Comment by sbwalker** [comment]
Actually the more important topic will be how to deal with .NET 5. The release of ,NET 5 in November will be a GA release - not an LTS release. Based on my experiences in my day job, enterprise customers are avoiding GA releases and sticking to LTS releases because with LTS they have guaranteed support and less risk. However the LTS release for .NET 5 is not planned until Nov 2021.... and the last LTS release was .NET Core 3.1 in Dec 2019... so there is a big gap between LTS releases. This is further complicated by the fact that what I just mentioned is true for all of the .NET stack EXCEPT for the Blazor WebAssembly bits ( excluding Blazor Server ) which were only GA in May - not LTS.

**Reply by PizzaConsole** [reply]
While I agree with your assessment of LTS. Microsoft has talked a big game about .NET 5 and it looks to be very promising. I am very eager to get my hands on it. Since this is a community of devs, I would at the least like to see a 'Dev' branch created for Oqtane when .NET 5 releases so that we can start working on the migration.

**Comment by sbwalker** [comment]
I think most folks will agree. There is definitely a challenge in maintaining 2 branches with a small community of developers. And I also think that custom modules which were compiled on .NET Core 3.1 may not run on .NET 5 ( and vice versa ). So there is a real impact to the developer community.

**Reply by PizzaConsole** [reply]
Very true, there is a lot to consider. Perhaps ideally we only migrate once we meet certain criteria, such as:

1. Necessary feature implementation
2. Security upgrade
3. Performance enhancement
4. Community demand

**Comment by kevind2000** [comment]
I'm one of the folks who would agree that development should be marching on the same platform version, currently it's 3.2, in the future it would be .NET 5 or 2021 LTS. After .NET 5 release, the immediate question might be whether the development should be migrate to .NET 5 early or later or even skip it?

**Reply by thabaum** [reply]
My thoughts was with my other discussion on when Oqtane v.Next or v2 would be released, which I was under the impression trying to time it with .NET 5.  So I noticed no replies there... so it must be way out there.  

Maybe v2 should be released with the LTS version?  That would give about 1 year to put out some more versions of 1 and allow us to play with the GA features.

I am in it for v2, so I am patient here as v1 I feel is ahead of it's time still.  More a look into the future of strictly Blazor developing using a modular framework today.

Shaun brings up some valuable points as I am more interested in the "Enterprise" side of things with LTS versions being something I would be more interested in investing into for more long term.  

Now Blazor has assembly lazy loading features along with more and more stuff each release like the new label component it is hard to want to do too much more except fine tune a few things along the way until a more concrete version of Blazor is released that incorporates all the fun stuff we as developers get to play with.

2 months ago you could not use these features, now they are here!

This is good food for thought thank you all for  sharing.

**Reply by PizzaConsole** [reply]
I think you make a good point. Now that Oqtane is in a rather stable place I do not see a reason to try and push too many new things into it until qe get .NET and can plan a solid migration path. I am still working on a few custom modules so thats soaks most of my time, as One of them is an enterprise application.

**Comment by kevind2000** [comment]
If Oqtane.vNext or v2 timed with .NET 5, does it mean the release also includes .NET 5 support, i.e., migration to .NET 5? If this community could support .NET 5 as soon as it released, it would be awesome from end user/early adaptor's point of view. On development side, Oqtane could take advantage of C# 9 and new features of Blazor and .NET 5, of course, not all of them, as Shaun mentioned in different post regarding to .NET 5 Preview 8. Early adoption of .NET 5 also means less gap to .NET 2021 LTS version.

Agree that Oqtane major/stable release as a product should be tied to .NET LTS version, given Oqtane is an innovative framework, it's worthwhile to move fast, i.e. to support .NET 5 early and let v2 or v3 aligned up with .NET LST version.  

Looking ahead something like:
- Oqtane v2 - .NET 5
- Oqtane v3 - .NET 6 LTS
- Oqtane v4 - .NET 7
- Oqtane v5 - .NET 8 LTS

**Comment by sbwalker** [comment]
The other major factor to consider is if we will lose any functionality by moving to .NET 5. In particular, I am concerned about the ability to support both runtime hosting models in the same application. Currently we have an excellent story around this, as it is controlled by a single appsettings.json entry. In .NET 5 I noticed that there are new elements in the *.csproj which would seem to make this impossible.

**Reply by PizzaConsole** [reply]
Nothing is truly impossible ;) (just could take too long to get right) In all seriousness I hope to spend more time soon looking into .NET 5 and see how it might affect Oqtane and better give my 2 cents worth

**Comment by sbwalker** [comment]
I exchanged emails with Dan Roth today and it sounds like because Oqtane is already shipping with dependencies on .NET Core 3.2 packages we are essentially already on a "Current" release – not an "LTS" release. And based on the .NET Core support model, what this means is that once .NET 5 ships, we will need to upgrade to .NET 5 within three months after .NET 5 ships or else we will be on an unsupported version of .NET Core. In addition, since the ability to use either Blazor Server or Blazor WebAssembly is a key value proposition for Oqtane then migrating to .NET 5 appears to be the right path forward – even if it is not an "LTS" release.

**Reply by kevind2000** [reply]
Sounds like a marching order! Look forward to contributing a little ...

**Reply by PizzaConsole** [reply]
I'd better start looking into .NET 5 sooner than later I suppose

**Reply by jiggyswift** [reply]
But irrespective of this, it simply makes sense to switch fast because Oqtane is such a young product.  It will be better for the longer term future of Oqtane to make the switch quickly.  I suspect people using Oqtane this early, are already quite early adopters.

**Comment by sbwalker** [comment]
We will need to test this but Dan seems to think that the Blazor WebAssembly packages in .NET 5 should be backwards compatible with Blazor WebAssembly 3.2 - so the transition for modules and themes may not be a breaking change

**Reply by PizzaConsole** [reply]
That would be fantastic news!
